### PR TITLE
245 delete stack

### PR DESCRIPTION
**Pull request by JohnCoene**
This PR was initiated to add the delete button to the stack #245 but I uncovered a few issues.

I don't think we have to "inject" the stack delete button, it should just be part of the UI.
The removal of stacks is handled at the workspace level, which seems strange to me; the stack "injects" the button in itself but the server code sits on the workspace server.
@DivadNojnarg may I change that? Is there something I am missing?

This PR starts this but is likely breaking things on the workspace server.

### CODE REVIEW

File: R/server.R [hunk 1]
```diff
@@ -359,6 +305,24 @@ generate_server.stack <- function(x, id = NULL, new_block = NULL,
...
 
+      observeEvent(input$remove, {
+        removeUI(
+          sprintf("#%s", session$ns(NULL))
+        )
+
+        rm_workspace_stack(id, workspace = workspace)
+      })
+
+      observeEvent(input$add, {
```
Comment (DivadNojnarg) [inline R/server.R:316]: What's the difference between 
```r
observeEvent(
        {
          req(new_block)
          new_block()
        },
        {
          add_block_stack(new_block()$block, new_block()$position, vals = vals)
        }
      )
```

and the `input$add` observer? Is it to handle the case where you have drag and drop?

File: R/server.R [hunk 2]
```diff
@@ -700,3 +663,63 @@ server_code.block <- function(x, state, output) {
     cat(deparse(generate_code(state())), sep = "\n")
   )
 }
+
+
+add_block_stack <- function(
```
Comment (DivadNojnarg) [inline R/server.R:668]: Thanks for reducing the observer code.

File: R/ui.R [hunk 3]
```diff
@@ -395,6 +388,14 @@ stack_header.stack <- function(x, title, ns, ...) {
...
         div(
           class = "stack-tools",
+          actionLink(
+            ns("remove"),
+            class = "text-decoration-none stack-remove",
+            `data-bs-toggle` = "tooltip",
+            `data-bs-title` = "Remove stack",
+            icon("trash")
+          ),
+          add_block_ui(ns),
```
Comment (DivadNojnarg) [inline R/ui.R:398]: Do you remove this piece of logic in the `blockr.demo`, as you're bringing blocks by drag and drop action?

**Review by DivadNojnarg — COMMENTED** [review]
I tested, the stack level function which seems to work fine. While quickly testing the workspace app, I end up with few issues: 
- Removing a stack does not update the number of stacks captured in `length(vals$stacks)`. 
- There're few floating tooltips remaining on the ui.
Screenshot 2024-02-16 at 15 36 46

File: R/ui.R (hunk 3 above)
Comment (JohnCoene) [reply R/ui.R:398]: I can always programmatically hide it from outside blockr if wanted. I don't see a problem with having both. 

I would like to improve this add block UI at a later stage after this is merged.

Comment (DivadNojnarg) [reply R/ui.R:398]: I am fine if it's not a problem for you. There are rooms for improvement regarding this piece of UI, it's pretty ugly ;). Feel free to improve in another PR.

**Comment by JohnCoene** [comment]
@nbenn last two commits fix the two remaining issues, good to review.

**Comment by JohnCoene** [comment]
Note on bbd4c183ac18404732abe1ef61be1e9c9a3c5482 I'm not sure it's best like that or to remove entirely.

**Comment by JohnCoene** [comment]
@DivadNojnarg @nbenn can we look into merging this?

**Comment by DivadNojnarg** [comment]
Will check later today

**Review by DivadNojnarg — COMMENTED** [review]
With the latest changes, tooltips correctly disappear. Minor comment: The layout is still quite strange. Feel free to fix if you can.
To reproduce:

```r
# library(blockr)
devtools::load_all()

do.call(set_workspace, args = c(list(), list(title = "My workspace")))

serve_workspace(clear = FALSE)
```

- Clear all stacks (button clear all)
- Add 3 empty stacks
- Remove the second stack: the third stacks remains in the third position, which leaves  a big gap.

**Comment by JohnCoene** [comment]
I pushed something to fix it.

**Review by DivadNojnarg — COMMENTED** [review]
There seem to be a remaining issue:

- Adding stacks 1 by 1: the first 2 stacks are vertically added and the other one are added horizontally to the right of the first stack (See video)

https://github.com/blockr-org/blockr/assets/18291543/dfb33d13-cf78-41fc-88ad-de693fa77f4d

**Comment by JohnCoene** [comment]
Alright, sorry about that. I fixed the issue and switched to using flex instead of Bootstrap's grid.

File: R/server.R [hunk 4]
```diff
@@ -442,24 +442,13 @@ generate_server.workspace <- function(x, id, ...) {
...
 
         stack_ui <- div(
-          class = "col stack-col",
+          class = "flex-grow-1 stack-col m-1",
           generate_ui(el, session$ns(stack_id))
         )
 
         insertUI(
-          selector = if (length(vals$stacks) == 0) {
-            ".workspace"
```
Comment (DivadNojnarg) [inline R/server.R:497]: @JohnCoene: removing this seems to break when we clear all stack and try to add a new stack. Nothing appear in the UI.

Comment (JohnCoene) [reply R/server.R:497]: Indeed, fixed in the clear stacks observer instead https://github.com/blockr-org/blockr/pull/298/commits/409918f2a9f8b5876513b71efb55904af6416ff1

**Review by DivadNojnarg — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**